### PR TITLE
Add elastic-net type costs with an orthogonal linear transformation of the input before it is fed into the regularizer

### PR DESCRIPTION
**Pull request by michalk8**
Implements ideas from [this paper](https://arxiv.org/abs/2306.11895). More specifically, this PR introduces a parameterized class of "elastic" translation invariant costs that can be described as sq. Euclidean distance + a regularizer applied to an orthogonal linear transform of the input. The MBO monge map estimator can be obtained by specifying the proximal operator of that regularizer.

TODOs:
- [x] fix the remaining TODOs (tau in sq-k-overlap, etc.)
- [x] prettify documentation + add the citation
- [x] check tutorial(s) to replace `gamma` with `scaling_reg`

### CODE REVIEW

File: src/ott/geometry/costs.py [hunk 1]
```diff
@@ -308,57 +307,168 @@ class RegTICost(TICost, abc.ABC):
   r"""Base class for regularized translation-invariant costs.
 
   .. math::
-
-    \frac{1}{2} \|\cdot\|_2^2 + reg\left(\cdot\right)
+    \frac{1}{2} \|\cdot\|_2^2 + \text{scaling_reg} reg\left(\cdot\right)
```
Comment (marcocuturi) [inline src/ott/geometry/costs.py:310]: `matrix` is missing?

File: src/ott/geometry/costs.py [hunk 2]
```diff
@@ -308,57 +307,168 @@ class RegTICost(TICost, abc.ABC):
...
-    self.gamma = gamma
+  def _reg(self, z: jnp.ndarray) -> float:  # noqa: D102
+    return 0.5 * jnp.sum(z ** 2)
+
+  def _prox_reg(self, z: jnp.ndarray, tau: float = 1.0) -> jnp.ndarray:
+    return z / (1.0 + tau * self.scaling_reg)
+
+
+@jax.tree_util.register_pytree_node_class
+class ElasticOrthL2(RegTICost):
```
Comment (marcocuturi) [inline src/ott/geometry/costs.py:454]: It feels this should be the same as `ElasticL2` with a matrix?

**Review by marcocuturi — COMMENTED** [review]
Thanks Michal! A few minor questions!

File: src/ott/geometry/costs.py [hunk 3]
```diff
@@ -47,8 +50,7 @@ class CostFn(abc.ABC):
   followed by a pairwise cost that involves both inputs, as in:
 
   .. math::
-
-    c(x,y) = norm(x) + norm(y) + pairwise(x,y)
+    c(x, y) = norm(x) + norm(y) + pairwise(x, y)
```
[the remark this answers is not in the document]
Comment (michalk8) [reply src/ott/geometry/costs.py:53]: No, it renders well!

File: src/ott/geometry/costs.py [hunk 4]
```diff
@@ -308,57 +307,168 @@ class RegTICost(TICost, abc.ABC):
...
+      \min_z h(z) - f(x - z). \\
+      \min_z h(z) + \tilde{f}(z, x).
+
+    where :math:`\tilde{f}(z, x) := -f(x - z)`.
+
+    TODO(michalk8): format this nicely
+    This is solved using proximal gradient descent, which requires having
+    access to the prox of `scaling_h * h` (and not only to that of `h`).
+    Given the properties of `h` (as squared-norm + `reg`), that prox is obtained
+    by rescaling the output of the prox of a suitable scaling of `reg`.
```
[the remark this answers is not in the document]
Comment (michalk8) [reply src/ott/geometry/costs.py:407]: It's already mentioned in the kwargs docstring.

File: src/ott/geometry/costs.py (hunk 1 above)
Comment (michalk8) [reply src/ott/geometry/costs.py:315]: Thanks, done!

**Pull request merged**